nvme.c

```diff
@@ -6674,6 +6679,177 @@ static int show_hostnqn_cmd(int argc, char **argv, struct command *command, stru
+		fprintf(stderr, "OpenSSL: initializing HMAC context failed\n");
+		err = -ENOENT;
+		goto out;
+	}
+
+	if (!HMAC_Update(hmac_ctx, (unsigned char *)hostnqn,
+			 strlen(hostnqn))) {
+		fprintf(stderr, "OpenSSL: HMAC for hostnqn failed\n");
+		err = -ENOENT;
+		goto out;
+	}
+
+	if (!HMAC_Update(hmac_ctx, (unsigned char *)hmac_seed,
+			 strlen(hmac_seed))) {
+		fprintf(stderr, "OpenSSL: HMAC for seed failed\n");
+		err = -ENOENT;
+		goto out;
+	}
+
+	if (!HMAC_Final(hmac_ctx, key, &key_len)) {
+		fprintf(stderr, "OpenSSL: finializing MAC failed\n");
+		err = -ENOENT;
+		goto out;
+	}
+
+	err = 0;
+
+out:
+	HMAC_CTX_free(hmac_ctx);
+	return err;
+}
+#endif /* !CONFIG_OPENSSL_1 */
+
+#ifdef CONFIG_OPENSSL_3
+static int gen_dhchap_key_ossl(char *hostnqn, unsigned int hmac,
+			       unsigned int key_len, unsigned char *secret,
+			       unsigned char *key)
+{
+	const char hmac_seed[] = "NVMe-over-Fabrics";
+	OSSL_PARAM params[2], *p = params;
+	OSSL_LIB_CTX *lib_ctx;
+	EVP_MAC_CTX *mac_ctx = NULL;
+	EVP_MAC *mac = NULL;
+	char *progq = NULL;
+	char *digest;
+	size_t len;
+	int err;
+
+	lib_ctx = OSSL_LIB_CTX_new();
+	if (!lib_ctx) {
+		fprintf(stderr, "OpenSSL: Library initializing failed\n");
+		return -ENOENT;
+	}
+
+	mac = EVP_MAC_fetch(lib_ctx, OSSL_MAC_NAME_HMAC, progq);
+	if (!mac) {
+		fprintf(stderr, "OpenSSL: could not fetch HMAC algorithm\n");
+		err = -EINVAL;
+		goto out;
+	}
+
+	mac_ctx = EVP_MAC_CTX_new(mac);
+	if (!mac_ctx) {
+		fprintf(stderr, "OpenSSL: could not create HMAC context\n");
+		err = -ENOENT;
+		goto out;
+	}
+
+	switch (hmac) {
+	case 1:
+		digest = OSSL_DIGEST_NAME_SHA2_256;
+		break;
+	case 2:
+		digest = OSSL_DIGEST_NAME_SHA2_384;
+		break;
+	case 3:
+		digest = OSSL_DIGEST_NAME_SHA2_512;
+		break;
+	default:
+		err = -EINVAL;
+		goto out;
+	}
+	*p++ = OSSL_PARAM_construct_utf8_string(OSSL_MAC_PARAM_DIGEST,
+						digest,
+						0);
+	*p = OSSL_PARAM_construct_end();
+
+	if (!EVP_MAC_init(mac_ctx, secret, key_len, params)) {
+		fprintf(stderr, "OpenSSL: could not initialize HMAC context\n");
+		err = -EINVAL;
+		goto out;
+	}
+
+	if (!EVP_MAC_update(mac_ctx, (unsigned char *)hostnqn,
+			    strlen(hostnqn))) {
+		fprintf(stderr, "OpenSSL: HMAC for hostnqn failed\n");
+		err = -ENOENT;
+		goto out;
+	}
+
+	if (!EVP_MAC_update(mac_ctx, (unsigned char *)hmac_seed,
+			    strlen(hmac_seed))) {
+		fprintf(stderr, "OpenSSL: HMAC for seed failed\n");
+		err = -ENOENT;
+		goto out;
+	}
+
+	if (!EVP_MAC_final(mac_ctx, key, &len, key_len)) {
+		fprintf(stderr, "OpenSSL: finializing MAC failed\n");
+		err = -ENOENT;
+		goto out;
+	}
+
+	if (len != key_len) {
+		fprintf(stderr, "OpenSSL: generated HMAC has an unexpected lenght\n");
+		err = -EINVAL;
+		goto out;
+	}
+
+	err = 0;
+
+out:
+	EVP_MAC_CTX_free(mac_ctx);
+	EVP_MAC_free(mac);
+	OSSL_LIB_CTX_free(lib_ctx);
+
+	return err;
+}
+#endif /* !CONFIG_OPENSSL_3 */
+
 static int gen_dhchap_key(int argc, char **argv, struct command *command, struct plugin *plugin)
```

Comment:
Here too.  Doesn't this belong in libnvme?

Reply:
Yep, I agree. I just factored this part out of the existing nvme-cli code base without thinking on moving it to the library.

Reply:
Done.